designs/defaults-and-model-evolution.md

```diff
@@ -0,0 +1,770 @@
+  `title` and `message` in code without having to first check if the value is
+  `null`. In many languages, one can call `message.message.size()` without
+  first checking if the value is non-optional.
+
+If the `title` member ever needs to be made optional, the `@required` trait
+can be replaced with the `@default` trait:
+
+```
+structure Message {
+    title: String = ""
+    message: String = ""
+}
+```
+
+With the above change, codegen remains the same: both values are non-optional.
+However, if `title` is not set in client code, server-side validation for the
+type will _not_ fail because `title` has a default value of `""`.
+
+
+## Proposal overview
+
+Introduce a 2.0 of Smithy models that simplifies optionality by moving
+optionality controls from shapes to structure members. Smithy IDL 2.0 will:
+
+1. Add a `@default` trait that can target structure members to assign a default
+   value. Whether a member has a default value is no longer controlled based on
+   the shape targeted by a member, localizing this concern to members. This
+   makes optionality of a member easier to understand for both readers and
+   writers.
+2. Add a `@clientOptional` trait that can target structure members. This trait
+   is essentially a more constrained and better named `@box` trait. The primary
+   use case for this trait is to apply it to members also marked as `@required`
+   to force non-authoritative code generators like clients to treat the member
+   as optional. The service reserves the right to remove the `@required` trait
+   from a member without replacing it with a `@default` trait without a major
+   version bump of the service.
+3. Remove the `@box` trait from the Smithy 2.0 prelude and fail to load models
+   that contain the `@box` trait.
+4. Remove the `PrimitiveBoolean`, `PrimitiveShort`, `PrimitiveInteger`,
+   `PrimitiveLong`, `PrimitiveFloat`, and `PrimitiveDouble` shapes from the
+   Smithy 2.0 prelude. IDL 2.0 models will fail if they use these shapes.
+5. Update the Smithy IDL 2.0 model loader implementation to be able to load
+   Smithy 1.0 models alongside Smithy 2.0 models.
+   1. Inject an appropriate `@default` trait on structure members that targeted
+      shapes with a default zero value and were not marked with the `@box`
+      trait.
+   2. Replace the `@box` trait with `@clientOptional` on structure members.
+   3. Remove the `@box` trait from non-members.
+   4. Rewrite members that target one of the removed Primitive* shapes to
+      target the corresponding non-primitive shape in the prelude (for example,
+      `PrimitiveInteger` is rewritten to target `Integer`).
+
+
+## `@default` trait
+
+The `@default` trait can be applied to structure members to provide a default
+value.
+
+The following example defines a structure with a "language" member that has a
+default value:
+
+```
+structure Message {
+    @required
+    title: String
+
+    language: Language = "en"
```

Comment:
Is there validation that asserts this is a real value?

Reply:
Yes, it's built-in via NodeValidationVisitor